File: NiftyOptionCalculator-iOS/NiftyOptionCalculator/ViewModels/TradeJournalViewModel.swift
import Foundation
import SwiftUI
import Combine
import os.log

private let logger = Logger(subsystem: "com.optix.app", category: "TradeJournalVM")

// MARK: - Journal Tab

enum JournalTab: String, CaseIterable {
    case entries = "Entries"
    case stats = "Stats"
}

// MARK: - Trade Journal ViewModel

@MainActor
final class TradeJournalViewModel: ObservableObject {

    // MARK: - Published Properties

    @Published var entries: [JournalEntry] = []
    @Published var stats: JournalStats?
    @Published var isLoading: Bool = false
    @Published var errorMessage: String?
    @Published var showError: Bool = false
    @Published var activeTab: JournalTab = .entries
    @Published var showForm: Bool = false
    @Published var editingEntry: JournalEntry?
    @Published var totalEntries: Int = 0

    // Filter state
    @Published var filterOutcome: TradeOutcome?
    @Published var filterSymbol: String = ""
    @Published var filterTag: JournalTag?

    // MARK: - Form Fields

    @Published var formSymbol: String = "NIFTY"
    @Published var formStrikePrice: String = ""
    @Published var formOptionType: String = "CE"
    @Published var formDirection: String = "buy"
    @Published var formEntryPrice: String = ""
    @Published var formExitPrice: String = ""
    @Published var formQuantity: String = "1"
    @Published var formLotSize: String = "75"
    @Published var formEntryDate: Date = Date()
    @Published var formExitDate: Date = Date()
    @Published var formHasExitDate: Bool = false
    @Published var formExpiryDate: Date = Date()
    @Published var formHasExpiryDate: Bool = true
    @Published var formTitle: String = ""
    @Published var formNotes: String = ""
    @Published var formSelectedTags: Set<JournalTag> = []
    @Published var formMood: TradeMood?
    @Published var formMarketCondition: MarketCondition?
    @Published var formOutcome: TradeOutcome?
    @Published var formPaperPositionId: String?

    // MARK: - Private Properties

    private let apiService = TradeJournalAPIService.shared
    private var cancellables = Set<AnyCancellable>()

    // Available symbols for the picker
    let availableSymbols = ["NIFTY", "BANKNIFTY", "FINNIFTY", "MIDCPNIFTY", "SENSEX", "BANKEX"]

    // Lot sizes per symbol
    let lotSizes: [String: Int] = [
        "NIFTY": 75,
        "BANKNIFTY": 30,
        "FINNIFTY": 25,
        "MIDCPNIFTY": 50,
        "SENSEX": 10,
        "BANKEX": 15
    ]

    // MARK: - Date Formatter

    private let dateFormatter: DateFormatter = {
        let f = DateFormatter()
        f.dateFormat = "yyyy-MM-dd"
        return f
    }()

    // MARK: - Initialization

    init() {
        // Auto-update lot size when symbol changes
        $formSymbol
            .sink { [weak self] symbol in
                if let lotSize = self?.lotSizes[symbol] {
                    self?.formLotSize = "\(lotSize)"
                }
            }
            .store(in: &cancellables)
    }

    // MARK: - Data Loading

    func loadEntries() async {
        guard AuthManager.shared.isLoggedIn else {
            entries = []
            return
        }

        isLoading = true
        defer { isLoading = false }

        do {
            let response = try await apiService.getEntries(
                outcome: filterOutcome?.rawValue,
                symbol: filterSymbol.isEmpty ? nil : filterSymbol,
                tag: filterTag?.rawValue,
                limit: 100
            )
            entries = response.entries
            totalEntries = response.total
            logger.info("Loaded \(response.entries.count) journal entries")
        } catch {
            logger.error("Failed to load entries: \(error.localizedDescription)")
            errorMessage = error.localizedDescription
            showError = true
        }
    }

    func loadStats() async {
        guard AuthManager.shared.isLoggedIn else {
            stats = nil
            return
        }

        isLoading = true
        defer { isLoading = false }

        do {
            stats = try await apiService.getStats()
            logger.info("Loaded journal stats")
        } catch {
            logger.error("Failed to load stats: \(error.localizedDescription)")
            errorMessage = error.localizedDescription
            showError = true
        }
    }

    // MARK: - CRUD Operations

    func createEntry() async {
        guard validateForm() else { return }

        isLoading = true
        defer { isLoading = false }

        let tagsString = formSelectedTags.isEmpty ? nil : formSelectedTags.map(\.rawValue).joined(separator: ",")

        let request = JournalCreateRequest(
            symbol: formSymbol,
            strike_price: Double(formStrikePrice) ?? 0,
            option_type: formOptionType,
            direction: formDirection,
            entry_price: Double(formEntryPrice) ?? 0,
            exit_price: formExitPrice.isEmpty ? nil : Double(formExitPrice),
            quantity: Int(formQuantity) ?? 1,
            lot_size: Int(formLotSize) ?? 75,
            entry_date: dateFormatter.string(from: formEntryDate),
            exit_date: formHasExitDate ? dateFormatter.string(from: formExitDate) : nil,
            expiry_date: formHasExpiryDate ? dateFormatter.string(from: formExpiryDate) : nil,
            title: formTitle.isEmpty ? nil : formTitle,
            notes: formNotes.isEmpty ? nil : formNotes,
            tags: tagsString,
            market_condition: formMarketCondition?.rawValue,
            mood: formMood?.rawValue,
            outcome: formOutcome?.rawValue,
            paper_position_id: formPaperPositionId
        )

        do {
            let entry = try await apiService.createEntry(request)
            entries.insert(entry, at: 0)
            totalEntries += 1
            showForm = false
            resetForm()
            logger.info("Created journal entry: \(entry.id)")
        } catch {
            logger.error("Failed to create entry: \(error.localizedDescription)")
            errorMessage = error.localizedDescription
            showError = true
        }
    }

    func updateEntry() async {
        guard let entry = editingEntry else { return }

        isLoading = true
        defer { isLoading = false }

        let tagsString = formSelectedTags.isEmpty ? nil : formSelectedTags.map(\.rawValue).joined(separator: ",")

        let request = JournalUpdateRequest(
            exit_price: formExitPrice.isEmpty ? nil : Double(formExitPrice),
            exit_date: formHasExitDate ? dateFormatter.string(from: formExitDate) : nil,
            title: formTitle.isEmpty ? nil : formTitle,
            notes: formNotes.isEmpty ? nil : formNotes,
            tags: tagsString,
            market_condition: formMarketCondition?.rawValue,
            mood: formMood?.rawValue,
            outcome: formOutcome?.rawValue
        )

        do {
            let updated = try await apiService.updateEntry(id: entry.id, request)
            if let index = entries.firstIndex(where: { $0.id == entry.id }) {
                entries[index] = updated
            }
            showForm = false
            editingEntry = nil
            resetForm()
            logger.info("Updated journal entry: \(entry.id)")
        } catch {
            logger.error("Failed to update entry: \(error.localizedDescription)")
            errorMessage = error.localizedDescription
            showError = true
        }
    }

    func deleteEntry(_ entry: JournalEntry) async {
        do {
            try await apiService.deleteEntry(id: entry.id)
            entries.removeAll { $0.id == entry.id }
            totalEntries -= 1
            logger.info("Deleted journal entry: \(entry.id)")
        } catch {
            logger.error("Failed to delete entry: \(error.localizedDescription)")
            errorMessage = error.localizedDescription
            showError = true
        }
    }

    // MARK: - Form Helpers

    func resetForm() {
        formSymbol = "NIFTY"
        formStrikePrice = ""
        formOptionType = "CE"
        formDirection = "buy"
        formEntryPrice = ""
        formExitPrice = ""
        formQuantity = "1"
        formLotSize = "75"
        formEntryDate = Date()
        formExitDate = Date()
        formHasExitDate = false
        formExpiryDate = Date()
        formHasExpiryDate = true
        formTitle = ""
        formNotes = ""
        formSelectedTags = []
        formMood = nil
        formMarketCondition = nil
        formOutcome = nil
        formPaperPositionId = nil
        editingEntry = nil
    }

    func populateFormForEditing(_ entry: JournalEntry) {
        editingEntry = entry
        formSymbol = entry.symbol
        formStrikePrice = String(format: "%.0f", entry.strike_price)
        formOptionType = entry.option_type
        formDirection = entry.direction
        formEntryPrice = String(format: "%.2f", entry.entry_price)
        formExitPrice = entry.exit_price.map { String(format: "%.2f", $0) } ?? ""
        formQuantity = "\(entry.quantity)"
        formLotSize = "\(entry.lot_size)"

        // Parse dates
        if let date = dateFormatter.date(from: entry.entry_date) {
            formEntryDate = date
        }
        if let exitDateStr = entry.exit_date, let date = dateFormatter.date(from: exitDateStr) {
            formExitDate = date
            formHasExitDate = true
        } else {
            formHasExitDate = false
        }
        if let expiryStr = entry.expiry_date, let date = dateFormatter.date(from: expiryStr) {
            formExpiryDate = date
            formHasExpiryDate = true
        } else {
            formHasExpiryDate = false
        }

        formTitle = entry.title ?? ""
        formNotes = entry.notes ?? ""

        // Parse tags
        if let tagsStr = entry.tags {
            formSelectedTags = Set(tagsStr.split(separator: ",").compactMap { JournalTag(rawValue: String($0).trimmingCharacters(in: .whitespaces)) })
        } else {
            formSelectedTags = []
        }

        formMood = entry.mood.flatMap { TradeMood(rawValue: $0) }
        formMarketCondition = entry.market_condition.flatMap { MarketCondition(rawValue: $0) }
        formOutcome = entry.outcome.flatMap { TradeOutcome(rawValue: $0) }
        formPaperPositionId = entry.paper_position_id

        showForm = true
    }

    func startNewEntry() {
        resetForm()
        editingEntry = nil
        showForm = true
    }

    // MARK: - Validation

    private func validateForm() -> Bool {
        guard !formStrikePrice.isEmpty, Double(formStrikePrice) != nil else {
            errorMessage = "Please enter a valid strike price"
            showError = true
            return false
        }
        guard !formEntryPrice.isEmpty, Double(formEntryPrice) != nil else {
            errorMessage = "Please enter a valid entry price"
            showError = true
            return false
        }
        guard !formQuantity.isEmpty, let qty = Int(formQuantity), qty > 0 else {
            errorMessage = "Please enter a valid quantity"
            showError = true
            return false
        }
        return true
    }

    // MARK: - Computed Helpers

    /// Compute P&L for an entry that has an exit price
    func computePnL(for entry: JournalEntry) -> Double? {
        guard let exitPrice = entry.exit_price else { return entry.realized_pnl }
        if let pnl = entry.realized_pnl { return pnl }
        let multiplier: Double = entry.direction == "buy" ? 1.0 : -1.0
        return (exitPrice - entry.entry_price) * multiplier * Double(entry.quantity) * Double(entry.lot_size)
    }

    func pnlColor(for entry: JournalEntry) -> Color {
        guard let pnl = computePnL(for: entry) else { return Theme.textMuted }
        if pnl > 0 { return Theme.profit }
        if pnl < 0 { return Theme.loss }
        return Theme.textMuted
    }
}

File: OptixAndroid/app/src/main/java/com/optix/app/presentation/screens/journal/TradeJournalViewModel.kt
package com.optix.app.presentation.screens.journal

import androidx.compose.ui.graphics.Color
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.optix.app.core.util.Resource
import com.optix.app.domain.model.*
import com.optix.app.domain.repository.AuthRepository
import com.optix.app.domain.repository.JournalRepository
import com.optix.app.presentation.theme.Loss
import com.optix.app.presentation.theme.Profit
import com.optix.app.presentation.theme.TextMutedDark
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import javax.inject.Inject
import kotlin.math.abs

// ============== UI State ==============

enum class JournalTab(val title: String) {
    ENTRIES("Entries"),
    STATS("Stats")
}

data class TradeJournalState(
    // Auth
    val isLoggedIn: Boolean = false,

    // Tab
    val activeTab: JournalTab = JournalTab.ENTRIES,

    // Entries
    val entries: List<JournalEntry> = emptyList(),
    val totalEntries: Int = 0,
    val isLoading: Boolean = false,
    val errorMessage: String? = null,

    // Stats
    val stats: JournalStats? = null,

    // Filters
    val filterOutcome: TradeOutcome? = null,
    val filterSymbol: String = "",
    val filterTag: JournalTag? = null,

    // Form visibility
    val showForm: Boolean = false,
    val editingEntry: JournalEntry? = null,

    // Delete confirmation
    val entryToDelete: JournalEntry? = null,

    // Form fields
    val formSymbol: String = "NIFTY",
    val formStrikePrice: String = "",
    val formOptionType: String = "CE",
    val formDirection: String = "buy",
    val formEntryPrice: String = "",
    val formExitPrice: String = "",
    val formQuantity: String = "1",
    val formLotSize: String = "75",
    val formEntryDate: Date = Date(),
    val formExitDate: Date = Date(),
    val formHasExitDate: Boolean = false,
    val formExpiryDate: Date = Date(),
    val formHasExpiryDate: Boolean = true,
    val formTitle: String = "",
    val formNotes: String = "",
    val formSelectedTags: Set<JournalTag> = emptySet(),
    val formMood: TradeMood? = null,
    val formMarketCondition: MarketCondition? = null,
    val formOutcome: TradeOutcome? = null
)

@HiltViewModel
class TradeJournalViewModel @Inject constructor(
    private val repository: JournalRepository,
    private val authRepository: AuthRepository
) : ViewModel() {

    companion object {
        val AVAILABLE_SYMBOLS = listOf("NIFTY", "BANKNIFTY", "FINNIFTY", "MIDCPNIFTY", "SENSEX", "BANKEX")
        val LOT_SIZES = mapOf(
            "NIFTY" to 75,
            "BANKNIFTY" to 30,
            "FINNIFTY" to 25,
            "MIDCPNIFTY" to 50,
            "SENSEX" to 10,
            "BANKEX" to 15
        )
    }

    private val _state = MutableStateFlow(TradeJournalState())
    val state: StateFlow<TradeJournalState> = _state.asStateFlow()

    private val dateFormatter = SimpleDateFormat("yyyy-MM-dd", Locale.US)

    init {
        observeAuthState()
    }

    private fun observeAuthState() {
        viewModelScope.launch {
            authRepository.authState.collect { authState ->
                val isLoggedIn = authState is AuthState.Authenticated
                _state.update { it.copy(isLoggedIn = isLoggedIn) }
                if (isLoggedIn) {
                    loadEntries()
                    loadStats()
                } else {
                    _state.update {
                        TradeJournalState(isLoggedIn = false)
                    }
                }
            }
        }
    }

    // ============== Data Loading ==============

    fun loadEntries() {
        viewModelScope.launch {
            _state.update { it.copy(isLoading = true, errorMessage = null) }
            val currentState = _state.value
            when (val result = repository.getEntries(
                outcome = currentState.filterOutcome?.apiValue,
                symbol = currentState.filterSymbol.ifEmpty { null },
                tag = currentState.filterTag?.apiValue,
                limit = 100
            )) {
                is Resource.Success -> {
                    val (entries, total) = result.data!!
                    _state.update {
                        it.copy(
                            entries = entries,
                            totalEntries = total,
                            isLoading = false
                        )
                    }
                }
                is Resource.Error -> {
                    _state.update {
                        it.copy(
                            errorMessage = result.message,
                            isLoading = false
                        )
                    }
                }
                is Resource.Loading -> {}
            }
        }
    }

    fun loadStats() {
        viewModelScope.launch {
            when (val result = repository.getStats()) {
                is Resource.Success -> {
                    _state.update { it.copy(stats = result.data) }
                }
                is Resource.Error -> {
                    _state.update { it.copy(errorMessage = result.message) }
                }
                is Resource.Loading -> {}
            }
        }
    }

    // ============== Tab ==============

    fun setActiveTab(tab: JournalTab) {
        _state.update { it.copy(activeTab = tab) }
    }

    // ============== Filters ==============

    fun setFilterOutcome(outcome: TradeOutcome?) {
        _state.update { it.copy(filterOutcome = outcome) }
        loadEntries()
    }

    fun setFilterSymbol(symbol: String) {
        _state.update { it.copy(filterSymbol = symbol) }
        loadEntries()
    }

    fun setFilterTag(tag: JournalTag?) {
        _state.update { it.copy(filterTag = tag) }
        loadEntries()
    }

    fun clearFilters() {
        _state.update {
            it.copy(filterOutcome = null, filterSymbol = "", filterTag = null)
        }
        loadEntries()
    }

    // ============== Form ==============

    fun startNewEntry() {
        _state.update {
            it.copy(
                showForm = true,
                editingEntry = null,
                formSymbol = "NIFTY",
                formStrikePrice = "",
                formOptionType = "CE",
                formDirection = "buy",
                formEntryPrice = "",
                formExitPrice = "",
                formQuantity = "1",
                formLotSize = "75",
                formEntryDate = Date(),
                formExitDate = Date(),
                formHasExitDate = false,
                formExpiryDate = Date(),
                formHasExpiryDate = true,
                formTitle = "",
                formNotes = "",
                formSelectedTags = emptySet(),
                formMood = null,
                formMarketCondition = null,
                formOutcome = null
            )
        }
    }

    fun startEditEntry(entry: JournalEntry) {
        _state.update {
            it.copy(
                showForm = true,
                editingEntry = entry,
                formSymbol = entry.symbol,
                formStrikePrice = String.format("%.0f", entry.strikePrice),
                formOptionType = entry.optionType,
                formDirection = entry.direction,
                formEntryPrice = String.format("%.2f", entry.entryPrice),
                formExitPrice = entry.exitPrice?.let { p -> String.format("%.2f", p) } ?: "",
                formQuantity = entry.quantity.toString(),
                formLotSize = entry.lotSize.toString(),
                formEntryDate = parseDate(entry.entryDate) ?: Date(),
                formExitDate = entry.exitDate?.let { parseDate(it) } ?: Date(),
                formHasExitDate = entry.exitDate != null,
                formExpiryDate = entry.expiryDate?.let { parseDate(it) } ?: Date(),
                formHasExpiryDate = entry.expiryDate != null,
                formTitle = entry.title ?: "",
                formNotes = entry.notes ?: "",
                formSelectedTags = entry.tags?.split(",")
                    ?.mapNotNull { JournalTag.fromApiValue(it.trim()) }
                    ?.toSet() ?: emptySet(),
                formMood = entry.mood?.let { TradeMood.fromApiValue(it) },
                formMarketCondition = entry.marketCondition?.let { MarketCondition.fromApiValue(it) },
                formOutcome = entry.outcome?.let { TradeOutcome.fromApiValue(it) }
            )
        }
    }

    fun dismissForm() {
        _state.update { it.copy(showForm = false, editingEntry = null) }
    }

    // Form field updaters
    fun updateFormSymbol(value: String) {
        val lotSize = LOT_SIZES[value] ?: 75
        _state.update { it.copy(formSymbol = value, formLotSize = lotSize.toString()) }
    }

    fun updateFormStrikePrice(value: String) = _state.update { it.copy(formStrikePrice = value) }
    fun updateFormOptionType(value: String) = _state.update { it.copy(formOptionType = value) }
    fun updateFormDirection(value: String) = _state.update { it.copy(formDirection = value) }
    fun updateFormEntryPrice(value: String) = _state.update { it.copy(formEntryPrice = value) }
    fun updateFormExitPrice(value: String) = _state.update { it.copy(formExitPrice = value) }
    fun updateFormQuantity(value: String) = _state.update { it.copy(formQuantity = value) }
    fun updateFormLotSize(value: String) = _state.update { it.copy(formLotSize = value) }
    fun updateFormEntryDate(value: Date) = _state.update { it.copy(formEntryDate = value) }
    fun updateFormExitDate(value: Date) = _state.update { it.copy(formExitDate = value) }
    fun updateFormHasExitDate(value: Boolean) = _state.update { it.copy(formHasExitDate = value) }
    fun updateFormTitle(value: String) = _state.update { it.copy(formTitle = value) }
    fun updateFormNotes(value: String) = _state.update { it.copy(formNotes = value) }
    fun updateFormMood(value: TradeMood?) = _state.update { it.copy(formMood = value) }
    fun updateFormMarketCondition(value: MarketCondition?) = _state.update { it.copy(formMarketCondition = value) }
    fun updateFormOutcome(value: TradeOutcome?) = _state.update { it.copy(formOutcome = value) }

    fun toggleFormTag(tag: JournalTag) {
        _state.update {
            val newTags = it.formSelectedTags.toMutableSet()
            if (newTags.contains(tag)) newTags.remove(tag) else newTags.add(tag)
            it.copy(formSelectedTags = newTags)
        }
    }

    // ============== CRUD ==============

    fun saveEntry() {
        val s = _state.value
        // Validate
        if (s.formStrikePrice.isEmpty() || s.formStrikePrice.toDoubleOrNull() == null) {
            _state.update { it.copy(errorMessage = "Please enter a valid strike price") }
            return
        }
        if (s.formEntryPrice.isEmpty() || s.formEntryPrice.toDoubleOrNull() == null) {
            _state.update { it.copy(errorMessage = "Please enter a valid entry price") }
            return
        }
        if (s.formQuantity.isEmpty() || (s.formQuantity.toIntOrNull() ?: 0) <= 0) {
            _state.update { it.copy(errorMessage = "Please enter a valid quantity") }
            return
        }

        val tagsString = s.formSelectedTags.takeIf { it.isNotEmpty() }
            ?.joinToString(",") { it.apiValue }

        viewModelScope.launch {
            _state.update { it.copy(isLoading = true, errorMessage = null) }

            if (s.editingEntry != null) {
                // Update
                when (val result = repository.updateEntry(
                    id = s.editingEntry.id,
                    exitPrice = s.formExitPrice.toDoubleOrNull(),
                    exitDate = if (s.formHasExitDate) dateFormatter.format(s.formExitDate) else null,
                    title = s.formTitle.ifEmpty { null },
                    notes = s.formNotes.ifEmpty { null },
                    tags = tagsString,
                    marketCondition = s.formMarketCondition?.apiValue,
                    mood = s.formMood?.apiValue,
                    outcome = s.formOutcome?.apiValue
                )) {
                    is Resource.Success -> {
                        val updated = result.data!!
                        _state.update { state ->
                            state.copy(
                                entries = state.entries.map { if (it.id == updated.id) updated else it },
                                showForm = false,
                                editingEntry = null,
                                isLoading = false
                            )
                        }
                    }
                    is Resource.Error -> {
                        _state.update { it.copy(errorMessage = result.message, isLoading = false) }
                    }
                    is Resource.Loading -> {}
                }
            } else {
                // Create
                when (val result = repository.createEntry(
                    symbol = s.formSymbol,
                    strikePrice = s.formStrikePrice.toDouble(),
                    optionType = s.formOptionType,
                    direction = s.formDirection,
                    entryPrice = s.formEntryPrice.toDouble(),
                    exitPrice = s.formExitPrice.toDoubleOrNull(),
                    quantity = s.formQuantity.toInt(),
                    lotSize = s.formLotSize.toIntOrNull() ?: 75,
                    entryDate = dateFormatter.format(s.formEntryDate),
                    exitDate = if (s.formHasExitDate) dateFormatter.format(s.formExitDate) else null,
                    expiryDate = if (s.formHasExpiryDate) dateFormatter.format(s.formExpiryDate) else null,
                    title = s.formTitle.ifEmpty { null },
                    notes = s.formNotes.ifEmpty { null },
                    tags = tagsString,
                    marketCondition = s.formMarketCondition?.apiValue,
                    mood = s.formMood?.apiValue,
                    outcome = s.formOutcome?.apiValue,
                    paperPositionId = null
                )) {
                    is Resource.Success -> {
                        val created = result.data!!
                        _state.update { state ->
                            state.copy(
                                entries = listOf(created) + state.entries,
                                totalEntries = state.totalEntries + 1,
                                showForm = false,
                                isLoading = false
                            )
                        }
                    }
                    is Resource.Error -> {
                        _state.update { it.copy(errorMessage = result.message, isLoading = false) }
                    }
                    is Resource.Loading -> {}
                }
            }
        }
    }

    fun requestDeleteEntry(entry: JournalEntry) {
        _state.update { it.copy(entryToDelete = entry) }
    }

    fun dismissDeleteConfirmation() {
        _state.update { it.copy(entryToDelete = null) }
    }

    fun confirmDeleteEntry() {
        val entry = _state.value.entryToDelete ?: return
        viewModelScope.launch {
            when (val result = repository.deleteEntry(entry.id)) {
                is Resource.Success -> {
                    _state.update { state ->
                        state.copy(
                            entries = state.entries.filter { it.id != entry.id },
                            totalEntries = state.totalEntries - 1,
                            entryToDelete = null
                        )
                    }
                }
                is Resource.Error -> {
                    _state.update { it.copy(errorMessage = result.message, entryToDelete = null) }
                }
                is Resource.Loading -> {}
            }
        }
    }

    fun dismissError() {
        _state.update { it.copy(errorMessage = null) }
    }

    // ============== Computed Helpers ==============

    fun computePnl(entry: JournalEntry): Double? {
        if (entry.realizedPnl != null) return entry.realizedPnl
        val exitPrice = entry.exitPrice ?: return null
        val multiplier = if (entry.direction == "buy") 1.0 else -1.0
        return (exitPrice - entry.entryPrice) * multiplier * entry.quantity * entry.lotSize
    }

    fun pnlColor(entry: JournalEntry): Color {
        val pnl = computePnl(entry) ?: return TextMutedDark
        return when {
            pnl > 0 -> Profit
            pnl < 0 -> Loss
            else -> TextMutedDark
        }
    }

    private fun parseDate(dateString: String): Date? {
        return try {
            dateFormatter.parse(dateString)
        } catch (e: Exception) {
            // Try ISO format
            try {
                SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss", Locale.US).parse(dateString)
            } catch (e2: Exception) {
                null
            }
        }
    }
}

/**
 * Format currency in Indian style
 */
fun formatIndianCurrency(value: Double): String {
    val absValue = abs(value)
    val sign = if (value < 0) "-" else if (value > 0) "+" else ""
    if (value == 0.0) return "\u20B90"
    return when {
        absValue >= 1_00_00_000 -> "$sign\u20B9${String.format("%.2f", absValue / 1_00_00_000)} Cr"
        absValue >= 1_00_000 -> "$sign\u20B9${String.format("%.2f", absValue / 1_00_000)}L"
        else -> "$sign\u20B9${String.format("%.0f", absValue)}"
    }
}
